src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs

```diff
@@ -92,14 +94,18 @@ private void Run()
             {
                 _delayStart.TrySetException(ex);
             }
+            _serviceStopped.TrySetResult(null);
         }
 
         public Task StopAsync(CancellationToken cancellationToken)
```

Comment:
Yes it's part of the interface https://github.com/dotnet/runtime/blob/967250cc86b4c613015066b47b882e7424fdb590/src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/IHostLifetime.cs#L27
Funny enough, none of the other implementations of this interface actually do anything in StopAsync so the cancellationToken is likely there because we asked for it in FxDC ;).

While we can't cancel the call to Stop, we could check before running it (throwing OCE right before the call to Stop by passing in the token), we could also cancel waiting for Run to complete.  That doesn't really cancel the work since the thread would still be active running the native service dispatcher loop.  We cannot "rollback" anything once the `Stop` method has been called - that's a one way operation.

What's our thinking on this - what's the right meaning for cancel on a Stop operation?

Reply:
> we could check before running it (throwing OCE right before the call to Stop by passing in the token), we could also cancel waiting for Run to complete

We should do that upfront check.  We could also pass the token to Task.Run, which will cause that task to be canceled if cancellation is requested before the delegate has been invoked.  Canceling the wait on that Task.Run (e.g. via a WaitAsync) is a larger leap, as it means you end up having work that's pending out there that's become disconnected from the caller's Task and they have no way to track it at that point; we should avoid doing that unless we have a very good reason. Of course, if you're about to tear down the process anyway, maybe it doesn't matter.

Reply:
Sounds good I made this change and added a test.

> Of course, if you're about to tear down the process anyway, maybe it doesn't matter.

I'm reluctant to do this.  Not allowing Run to exit gracefully means SCM may not treat the service as stopped, and if it's an auto-start service, it will *restart* it.